src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs

```diff
@@ -17,24 +18,45 @@ namespace Microsoft.AspNetCore.HttpSys.Internal
     internal unsafe class NativeRequestContext : IDisposable
     {
         private const int AlignmentPadding = 8;
+        private const int DefaultBufferSize = 4096 - AlignmentPadding;
         private IntPtr _originalBufferAddress;
         private HttpApiTypes.HTTP_REQUEST* _nativeRequest;
-        private byte[] _backingBuffer;
+        private IMemoryOwner<byte> _backingBuffer;
+        private MemoryHandle _memoryHandle;
         private int _bufferAlignment;
         private SafeNativeOverlapped _nativeOverlapped;
         private bool _permanentlyPinned;
+        private bool _disposed;
 
         // To be used by HttpSys
-        internal NativeRequestContext(SafeNativeOverlapped nativeOverlapped,
-            int bufferAlignment,
-            HttpApiTypes.HTTP_REQUEST* nativeRequest,
-            byte[] backingBuffer,
-            ulong requestId)
+        internal NativeRequestContext(SafeNativeOverlapped nativeOverlapped, MemoryPool<Byte> memoryPool, uint? bufferSize, ulong requestId)
         {
             _nativeOverlapped = nativeOverlapped;
-            _bufferAlignment = bufferAlignment;
-            _nativeRequest = nativeRequest;
-            _backingBuffer = backingBuffer;
+
+            // TODO:
+            // Apparently the HttpReceiveHttpRequest memory alignment requirements for non - ARM processors
+            // are different than for ARM processors. We have seen 4 - byte - aligned buffers allocated on
+            // virtual x64/x86 machines which were accepted by HttpReceiveHttpRequest without errors. In
+            // these cases the buffer alignment may cause reading values at invalid offset. Setting buffer
+            // alignment to 0 for now.
+            // 
+            // _bufferAlignment = (int)(requestAddress.ToInt64() & 0x07);
```

Comment:
So did HttpSysServer ever work on ARM? Would this regress that scenario?

Reply:
Not sure if it works at the moment, the requirements have shifted several times. Nothing has changed here for ARM, this code only got moved.